R/tab-book.R

```diff
@@ -112,6 +129,161 @@ tabBook <- function(multitable, dataset, weight = crunch::weight(dataset),
     }
 }
 
+#' @importFrom stats ave
+#' @importFrom utils stack
+tabBookMulti <- function(
+    multitable,
+    dataset,
+    weight_spec,
+    output_format,
+    file,
+    filter,
+    use_legacy_endpoint,
+    dots
+) {
+    if (length(weight_spec) == 0) {
+        stop("Empty list not allowed as a weight spec, use NULL to indicate no weights")
+    }
+
+    if (output_format != "json") {
+        stop("Complex weights only supported for json tabBooks")
+    }
+
+    if (is.data.frame(weight_spec) && !setequal(names(weight_spec), c("weight", "alias"))) {
+        stop("if weight_spec is a data.frame it must have exactly two columns: 'weight' & 'alias'")
+    }
+
+    if (!is.data.frame(weight_spec)) weight_spec <- tabBookWeightSpec(dataset, weight_spec)
+
+    if (any(duplicated(weight_spec))) {
+        stop("Found duplicate weight and alias combinations in weight_spec")
+    }
+
+    wt_vars <- unique(weight_spec$weight)
+    # Add a column that indicates what page the variable will be on
+    # in the weight-specific tabbook
+    weight_spec$page_num <- as.numeric(ave(weight_spec$weight, weight_spec$weight, FUN = seq_along))
+
+    books <- lapply(wt_vars, function(wt) {
+        page_vars <- weight_spec$alias[weight_spec$weight == wt]
+
+        tabBookSingle(
+            multitable,
+            dataset[page_vars],
+            dataset[[wt]],
+            output_format,
+            file = NULL,
+            filter,
+            use_legacy_endpoint,
+            dots
+        )
+    })
+    names(books) <- wt_vars
+
+    # stitch together
+    # Most of the objects should be the same because they come from the same multitable
+    # But the analyses object contain an item per variable with the weight included
+    # and then the pages section contains the cube results
+    analyses <- mapply(
+        weight = weight_spec$weight,
+        page_num = weight_spec$page_num,
+        FUN = function(weight, page_num) {
+            books[[which(names(books) == weight)]]@.Data[[1]]$analyses[[page_num]]
```

Comment:
This might need to be `[page_num]` in single square brackets to match the original format of a TabBookResult.

Reply:
Decided against this, this way it better matches single tabBooks to have double.